packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,182 @@
+      }
+      this.landmarks.splice(insertPosition, 0, {ref, role, label});
+    }
+  }
+
+  public removeLandmark(ref: MutableRefObject<HTMLElement>) {
+    this.landmarks = this.landmarks.filter(landmark => landmark.ref !== ref);
+  }
+
+  // Gets the landmark that is the closest parent in the DOM to the child
+  // Useful for nested Landmarks
+  private closestLandmark(element: HTMLElement) {
+    let landmarkMap = new Map(this.landmarks.map(l => [l.ref.current, l]));
+    let currentElement = element;
+    while (!landmarkMap.has(currentElement) && currentElement !== document.body) {
+      currentElement = currentElement.parentElement;
+    }
+    return landmarkMap.get(currentElement);
+  }
+
+  public getNextLandmark(element: HTMLElement, {backward}: {backward?: boolean }) {
+    let currentLandmark = this.closestLandmark(element);
+    let nextLandmarkIndex = 0;
+    if (currentLandmark) {
+      nextLandmarkIndex = this.landmarks.findIndex(landmark => landmark === currentLandmark) + (backward ? -1 : 1);
+    }
+
+    // Wrap if necessary
+    if (nextLandmarkIndex < 0) {
+      nextLandmarkIndex = this.landmarks.length - 1;
+    } else if (nextLandmarkIndex >= this.landmarks.length) {
+      nextLandmarkIndex = 0;
+    }
+
+    return this.landmarks[nextLandmarkIndex];
+  }
+
+  // skipping nodes in the current landmark, look for the next landmark, wrap if needed
+  // if that landmark had something focused in it previously, focus that
+  // else focus the first focusable element
+  // hopefully getFocusableTreeWalker will do the searching, or some derivative of it
+  // can we use FocusScope to handle the refocusing inside the landmark?
+  // essentially treat landmarks as a focus scope but without wrapping or restore or containment
+  public f6Handler(e: KeyboardEvent) {
+    if (e.key === 'F6') {
+      e.preventDefault();
+      e.stopPropagation();
+
+      let nextLandmark = this.getNextLandmark(e.target as HTMLElement, {backward: e.shiftKey});
+      
+      // If something was previously focused in the next landmark, then return focus to it
+      if (nextLandmark.lastFocused) {
+        let lastFocused = nextLandmark.lastFocused;
+        if (document.body.contains(lastFocused)) {
+          lastFocused.focus();
+          return;
+        }
+      }
+
+      // Otherwise, focus the first focusable element in the next landmark
+      let walker = getFocusableTreeWalker(nextLandmark.ref.current, {tabbable: true});
+      let nextNode = walker.nextNode() as HTMLElement;
+      if (!nextNode) {
+        nextNode = walker.firstChild() as HTMLElement;
+      }
+      nextNode.focus();
+    }
+  }
+
+  // Only the closest landmark cares about focus inside of it
+  public focusinHandler(e: FocusEvent) {
+    let currentLandmark = this.closestLandmark(e.target as HTMLElement);
+    this.landmarks = this.landmarks.map((landmark) => ({...landmark, ...(landmark === currentLandmark && {lastFocused: e.target as HTMLElement})}));
```

Comment:
lets add two functions

```
public updateLandmark(landmark: Landmark) {
  let roles = [];
  this.landmarks = this.landmarks.map((prevLandmark) =>  prevLandmark.ref === landmark ? {...prevLandmark, ...landmark} : prevLandmark);
  this.checkRolesAndLabels();
}
```
```
private checkRolesAndLabels() {
  // check landmarks in this.landmarks for role collisions and warn for any without labels
}
```

Now focusing can call update, and the useLandmark can more easily call things in a separation of concerns, I'll write a comment to detail that a bit more

Reply:
I'll probably do this by role as `checkLabels(role)` instead, so we aren't checking every landmark when just one changed.